File: backend/app/models/workflow.py
"""运维工作流(流程)模型 (D5 后端化: 流程数据由前端 localStorage 迁移至服务端)。

对应前端 WorkflowCenter 的 WorkflowItem:
  - approval / logs / knowledge_links 为嵌套结构, 用 JSON 列存储。
"""
from __future__ import annotations

import datetime

from sqlalchemy import JSON, Column, Integer, String, Text

from app.db.session import Base


def _now() -> str:
    return datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%d %H:%M:%S")


class WorkflowItem(Base):
    __tablename__ = "workflow_item"

    id = Column(String(32), primary_key=True)                 # INC-2026-0001
    type = Column(String(16), default="incident")              # incident/problem/change/risk
    title = Column(String(255), default="")
    description = Column(Text, default="")
    priority = Column(String(8), default="P3")                 # P1-P4
    status = Column(String(16), default="new")                 # new/progress/approval/rejected/closed
    owner = Column(String(64), default="")
    applicant = Column(String(64), default="")
    sla_hours = Column(Integer, default=24)
    risk_level = Column(String(16), default=None)              # high/medium/low (risk 类型)
    approval = Column(JSON, default=list)                      # [{approver,status,comment,at}]
    logs = Column(JSON, default=list)                          # [{user,text,at}]
    knowledge_links = Column(JSON, default=list)               # [kb_code]
    created_at = Column(String(32), default=_now)
    updated_at = Column(String(32), default=_now, onupdate=_now)

File: frontend/src/api/workflow.ts
import request from '@/api/request'

export type WType = 'incident' | 'problem' | 'change' | 'risk'
export type WStatus = 'new' | 'progress' | 'approval' | 'approved' | 'rejected' | 'closed'
export type WPriority = 'P1' | 'P2' | 'P3' | 'P4'
export type NodeStatus = 'approved' | 'rejected' | 'pending' | 'skipped'

export interface WNode {
  approver: string
  status: NodeStatus
  comment?: string
  at?: string
}
export interface WLog {
  user: string
  text: string
  at: string
}
export interface WorkflowItem {
  id: string
  type: WType
  title: string
  description: string
  priority: WPriority
  status: WStatus
  owner: string
  applicant: string
  createdAt: string
  updatedAt: string
  slaHours: number
  riskLevel?: 'high' | 'medium' | 'low'
  approval: WNode[]
  logs: WLog[]
  knowledgeLinks: string[]
}

export interface WorkflowListResp {
  items: WorkflowItem[]
}

export interface WorkflowTrendPoint {
  /** 该周周一, MM-DD */
  week: string
  created: number
  closed: number
}
export interface WorkflowStats {
  total: number
  open: number
  monthCreated: number
  monthClosed: number
  /** 平均解决时长(小时), 无已关闭流程时为 0 */
  avgResolve: number
  /** SLA 超时率百分比 (0-100) */
  breachRate: number
  byType: Record<string, number>
  byStatus: Record<string, number>
  byPriority: Record<string, number>
  trend: WorkflowTrendPoint[]
}

export const getWorkflows = () => request.get<unknown, WorkflowListResp>('/api/ops/workflows')
/** 服务端聚合统计(KPI + 分布 + 近 12 周趋势), 前端失败时回退本地计算 */
export const getWorkflowStats = () =>
  request.get<unknown, WorkflowStats>('/api/ops/workflows/stats')
export const createWorkflow = (data: Partial<WorkflowItem>) =>
  request.post<unknown, WorkflowItem>('/api/ops/workflows', data)
export const getWorkflow = (id: string) =>
  request.get<unknown, WorkflowItem>(`/api/ops/workflows/${id}`)
export const updateWorkflow = (id: string, patch: Partial<WorkflowItem>) =>
  request.put<unknown, WorkflowItem>(`/api/ops/workflows/${id}`, patch)
export const deleteWorkflow = (id: string) =>
  request.delete<unknown, void>(`/api/ops/workflows/${id}`)
export const advanceWorkflow = (id: string) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/advance`)
export const closeWorkflow = (id: string) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/close`)
export const reopenWorkflow = (id: string) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/reopen`)
export const approveNode = (
  id: string,
  nodeIndex: number,
  result: 'approved' | 'rejected',
  comment?: string,
) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/approve`, {
    node_index: nodeIndex,
    result,
    comment,
  })
export const addWorkflowLog = (id: string, text: string) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/logs`, { text })
export const linkKnowledge = (id: string, kbId: string) =>
  request.post<unknown, WorkflowItem>(`/api/ops/workflows/${id}/link`, { kb_id: kbId })
export const unlinkKnowledge = (id: string, kbId: string) =>
  request.delete<unknown, WorkflowItem>(`/api/ops/workflows/${id}/link/${kbId}`)
